8952 - bug - serve documents

matthopson opened the pull request:
Uses case status to determine if documents are eligible for service on a case.

matthopson committed:
8952 - Case Entity - method to check if document service is allowed

matthopson committed:
8952 - 8763 - expose utility for checking if case allows service

matthopson committed:
8952 - 8763 - refactor service allowance in paper docket helper

matthopson committed:
8952 - 8763 - refactor service allowance in add court issued docket entry

matthopson committed:
8952 - 8763 - refactor document viewer service button display

matthopson committed:
8952 - 8763 - refactor message document service button display

matthopson committed:
8952 - 8763 - update logic for petition service with case service allowance

matthopson committed:
Fix pa11y

matthopson committed:
8952 - 8763 - More refactors around document service allowance

matthopson committed:
8952 - fix test desc

matthopson committed:
8952 - 8763 - PublicCase Entity - Expose case status

matthopson commented:
Marking as draft, pending question in bug comment: https://github.com/flexion/ef-cms/issues/8952#issuecomment-924432911

matthopson commented:
Failing test re: the exposed case status.

mmarcotte reviewed (approved):
Looks super clean and clear

kkoskelin commented on shared/src/business/entities/cases/Case.js:
```diff
@@ -2267,8 +2267,20 @@ const caseHasServedDocketEntries = rawCase => {
+ * @returns {string} caseTitle the case caption with the postfix removed
+ */
+
+const canAllowDocumentServiceForCase = rawCase => {
```
While working on #7643 we created a function similar to this -- and I'm wondering if it should be used instead, or if its conditions are entirely different:
https://github.com/flexion/ef-cms/blob/7643-sqs-integration/shared/src/business/entities/cases/Case.js#L1193
(the above branch is part of code that is in a long-running story)
As you can see, our code also checks if case has been closed inside the last six months.  I'm not certain whether this is the way to do it, or if these require different logic.  Just wanted to put this out there.

matthopson replied:
I don't think that was the intent of this particular story. Per the PO, we want to disallow document service on a case that is in NEW status. This method is _currently_ only checking for that. I left it as a method instead of just checking its status everywhere in case this logic changed, at least tit would be abstracted. I'm wondering if I can just use the `isSttatusNew` property instead of exposing the status on Public case. That would give us what we want and we wouldn't have to make any changes to the entity, API response, or associated tests.

ujahio committed:
8952: correcting case tests [skip ci]

ujahio committed:
Merge branch '8952-bug-serve-documents' of https://github.com/flexion/ef-cms into 8952-bug-serve-documents

kkoskelin committed:
8952: fixing tests to conform to new behavior

updating tests on v1 and v2 getCaseLambda which now reveal more contact information

kkoskelin commented:
@mmarcotte heads-up that I have attempted to update some of the failing tests in this PR, but also there are some very slight changes to the `getCaseLambda` endpoints on v1 and v2.  You can see those changes reflected in the test updates I have committed.  I know that alterations to v1 and v2 are something not to be taken lightly, and I would like to discuss this with you.  There are no properties being added or removed from a potential response, but there is now some data being returned which was previously absent.

Happy to discuss in person!

mmarcotte commented on web-api/src/v1/getCaseLambda.test.js:
```diff
@@ -96,16 +96,17 @@ describe('getCaseLambda (which fails if version increase is needed, DO NOT CHANG
+      state: 'TN',
+    });
+    expect(parsedResponse.noticeOfTrialDate).toBeUndefined();
+    expect(parsedResponse.status).toBe('Calendared');
```
I think this part confuses me. the v1 API would only be consumed by the IRS Super User. And the IRS Super User would be associated with any case that is not in New Status.

mmarcotte replied:
Also, we should not be changing these tests. These are what keep the API returning what we have promised to the V1 and V2 API consumers.

mmarcotte commented on web-api/src/v2/getCaseLambda.test.js:
```diff
@@ -119,16 +119,17 @@ describe('getCaseLambda (which fails if version increase is needed, DO NOT CHANG
-    expect(JSON.parse(response.body).trialDate).toBeUndefined();
-    expect(JSON.parse(response.body).trialLocation).toBeUndefined();
-    expect(JSON.parse(response.body).userId).toBeUndefined();
+    const parsedResponse = JSON.parse(response.body);
+    expect(parsedResponse).toHaveProperty('caseCaption', expect.any(String));
+    expect(parsedResponse.assignedJudge).toBeUndefined();
+    expect(parsedResponse.contactPrimary).toMatchObject({
+      name: 'Test Petitioner',
+      state: 'TN',
+    });
+    expect(parsedResponse.status).toBe('Calendared');
+    expect(parsedResponse.trialDate).toBeUndefined();
+    expect(parsedResponse.trialLocation).toBeUndefined();
+    expect(parsedResponse.userId).toBeUndefined();
```
We should not be changing these tests. These are what keep the API returning what we have promised to the V1 and V2 API consumers.

kkoskelin committed:
8952: fixing v1 and v2 endpoint test expectations

status now allows contact information to also be seen

kkoskelin committed:
8952: coerce expression to boolean

fixing tests

kkoskelin committed:
8952: fixing seed data for cypress and pa11y test

kkoskelin committed:
8952: different case status in seed data

mmarcotte commented on web-api/src/v1/getCaseLambda.test.js:
```diff
@@ -96,16 +96,15 @@ describe('getCaseLambda (which fails if version increase is needed, DO NOT CHANG
+    expect(parsedResponse.trialLocation).toBeUndefined();
+    expect(parsedResponse.userId).toBeUndefined();
+    // contact and status is unavailable for user unassociated with the case
+    expect(parsedResponse.contactPrimary).toBeDefined();
+    expect(parsedResponse.status).toBeDefined();
```
these should be undefined as per the earlier tests.

adunkman replied:
As per discussion this morning, the `status` here was unintentionally missing from this API endpoint. It is included in the API contract and was set on some cases, but not always. This change is because for this specific test case, the case status _should have been_ and now _will_ be included. 

@kkoskelin, can you explain why `contactPrimary` is included as well? We talked about it this morning, and you mentioned that it was also not set sometimes when it should have been — but I can’t follow why it would have been missing before and is now set. Is there some filter which checks on `status`? 

In any case, the comment here is incorrect and contradicts the test assertions — probably worth dropping the comment so it doesn’t become outdated again.

kkoskelin replied:
@adunkman The particular case used in this test was _sealed_, and previously had no `status` value.  It means that the marshalling function attempted to obtain the `contactPrimary` as seen here:
https://github.com/ustaxcourt/ef-cms/pull/1412/files#diff-384b0330bd6db4b05ec5bf32f098eb0c01feab98b385ef5653e6dbc59faf58f8R1603

However, the "else" block executed wasn't using the correct method of obtaining the contact info because this case was *not* actually New -- it was Calendared.  The contact information was most definitely available on the PublicCase entity instance, but our code was doing a poor job of attempting to retrieve it, and thus it previously had been undefined.

mmarcotte commented on web-api/src/v2/getCaseLambda.test.js:
```diff
@@ -119,16 +119,14 @@ describe('getCaseLambda (which fails if version increase is needed, DO NOT CHANG
+    const parsedResponse = JSON.parse(response.body);
+    expect(parsedResponse).toHaveProperty('caseCaption', expect.any(String));
+    expect(parsedResponse.assignedJudge).toBeUndefined();
+    expect(parsedResponse.contactPrimary).toBeDefined();
+    expect(parsedResponse.status).toBeDefined();
```
these should be undefined per the earlier tests.

adunkman replied:
Same as above — will look to the above comment thread for response.

adunkman reviewed (commented):
Thanks for the explanation today, y’all!

kkoskelin committed:
8952: updating cypress test

[skip ci]

kkoskelin committed:
8952: remove unnecessary tab-click command

kkoskelin committed:
8952: fixing pa11y test reference

use a not-closed case for this test

mrinsin commented on shared/src/business/entities/cases/PublicCase.js:
```diff
@@ -104,6 +105,12 @@ const publicCaseSchema = {
+    ...Object.values(CASE_STATUS_TYPES),
+  )
+    .optional()
+    .meta({ tags: ['Restricted'] })
```
Hello! Just curious to know what this `[Restricted]` tag does!

kkoskelin replied:
@matthopson can you answer this?

matthopson replied:
This was pulled over from the other definitions of this field on Case. I'm not 100% sure, as I couldn't find where it was being used, but I thought these would be restricted fields on a blocked case - which makes this a good question to try and answer for sure.

matthopson replied:
Eric said it was used for document generation and doesn't change anything. :)

matthopson commented:
👍 (I'm the author - can't thumb this) - but thanks @kkoskelin

ericsorenson committed:
8952 add test data for MISC petitions [skip ci]

mrinsin commented on web-client/src/presenter/computeds/documentViewerHelper.test.js:
```diff
@@ -200,6 +203,28 @@ describe('documentViewerHelper', () => {
+      expect(showUnservedPetitionWarning).toBe(false);
+    });
+
+    it('should be true if an otherwise servable document is selected but the case is new', () => {
```
These are some great test descriptions! - Rachel, Rosie and Mrin!

adunkman commented on web-api/src/v1/getCaseLambda.test.js:
```diff
@@ -96,16 +96,15 @@ describe('getCaseLambda (which fails if version increase is needed, DO NOT CHANG
+    expect(parsedResponse.noticeOfTrialDate).toBeUndefined();
+    expect(parsedResponse.trialLocation).toBeUndefined();
+    expect(parsedResponse.userId).toBeUndefined();
+    // contact and status is unavailable for user unassociated with the case
```
```suggestion

```

adunkman reviewed (approved):
(non-blocking) Reminder that there’s a comment here which isn’t accurate, as previously mentioned.

kkoskelin committed:
Update web-api/src/v1/getCaseLambda.test.js

removing outdated comment

Co-authored-by: Andrew Dunkman (he/him) <adunkman@gmail.com>
